### PR TITLE
Add QGaussRadau quadrature formula + tests

### PR DESCRIPTION
**Pull request by jpthiele**
This adds the Gauss-Radau quadrature formulae up to n=3 points,
as described here: https://mathworld.wolfram.com/RadauQuadrature.html.
The given numerical values for n=4 and n=5 are not accurate enough
and application to the test polynomials
of quadrature_test.cc were not exact for degrees > 0.

On the 3d faces and subfaces the QF is able to integrate the given
polynomial in the test suite of one order higher than in all other cases.
This is reflected in the quadrature_test.ouput file for numdiff comparison.

### CODE REVIEW

**Comment by jpthiele** [comment]
Addendum: As the right Gauss-Radau formula quadrature points are obtained by subtracting the left quadrature points from 1,
the resulting formula is not sorted. This results in some additional output in the test file. 
I asked @kronbichler today whether the points should be reordered and he said that it should not matter to his knowledge,
so I opted for skipping a sorting step.

File: include/deal.II/base/quadrature_lib.h [hunk 1]
```diff
@@ -47,6 +47,58 @@ class QGauss : public Quadrature<dim>
...
+ * @sa https://mathworld.wolfram.com/RadauQuadrature.html @sa
+ */
+template <int dim>
+class QGaussRadau : public Quadrature<dim>
+{
+public:
+  /* EndPoint is used to specify which of the two endpoints of the unit interval
+   * is used also as quadrature point
+   */
+  enum EndPoint
```
Comment (peterrum) [inline include/deal.II/base/quadrature_lib.h:79]:
```suggestion
  enum class EndPoint
```

File: include/deal.II/base/quadrature_lib.h [hunk 2]
```diff
@@ -47,6 +47,58 @@ class QGauss : public Quadrature<dim>
...
+  QGaussRadau(const unsigned int n, EndPoint ep = QGaussRadau::left);
+
+  /**
+   * Move constructor. We cannot rely on the move constructor for `Quadrature`,
+   * since it does not know about the additional member `ep` of this class.
+   */
+  QGaussRadau(QGaussRadau<dim> &&) noexcept = default;
+
+private:
+  const EndPoint ep;
```
Comment (peterrum) [inline include/deal.II/base/quadrature_lib.h:99]:
```suggestion
  const EndPoint end_point;
```

File: include/deal.II/base/quadrature_lib.h [hunk 3]
```diff
@@ -47,6 +47,58 @@ class QGauss : public Quadrature<dim>
   QGauss(const unsigned int n);
 };
 
+/**
+ * The Gauss-Radau family of quadrature rules for numerical integration.
```
Comment (peterrum) [inline include/deal.II/base/quadrature_lib.h:51]: Could you add a sentence about when this quadrature rule is useful.

File: source/base/quadrature_lib.cc [hunk 4]
```diff
@@ -144,6 +144,113 @@ namespace internal
...
+      std::vector<double> points(n);
+      std::vector<double> left_points = get_left_quadrature_points(n);
+      switch (ep)
+        {
+          case ::dealii::QGaussRadau<1>::left:
+            return left_points;
+          case ::dealii::QGaussRadau<1>::right:
+            for (unsigned int i = 0; i < n; ++i)
+              {
+                points[i] = 1. - left_points[i];
```
Comment (peterrum) [inline source/base/quadrature_lib.cc:192]: Personally, I would prefer if the points were sorted. Could you flip the direction of the for loop?

File: source/base/quadrature_lib.cc [hunk 5]
```diff
@@ -144,6 +144,113 @@ namespace internal
...
+                "This constructor can only be called with either "
+                "QGaussRadau::left or QGaussRadau::right as second argument."));
+        }
+    }
+
+    // Implementes lookup table after affine transformation (*0.5)
+    // original for [-1,1] listed on
+    // https://mathworld.wolfram.com/RadauQuadrature.html
+    std::vector<double>
+    get_quadrature_weights(const unsigned int n)
```
Comment (peterrum) [inline source/base/quadrature_lib.cc:208]:
```suggestion
    get_left_quadrature_weights(const unsigned int n)
```

**Review by peterrum — COMMENTED** [review]
Looks overall good. I have some minor comments. Could you add a change-log entry?

File: source/base/quadrature_lib.cc [hunk 6]
```diff
@@ -144,6 +144,113 @@ namespace internal
...
+                points[i] = 1. - left_points[i];
+              }
+            return points;
+          default:
+            Assert(
+              false,
+              ExcMessage(
+                "This constructor can only be called with either "
+                "QGaussRadau::left or QGaussRadau::right as second argument."));
+        }
```
Comment (peterrum) [inline source/base/quadrature_lib.cc:201]:
```suggestion
          return {};
        }
```

File: include/deal.II/base/quadrature_lib.h (hunk 2 above)
Comment (jpthiele) [reply include/deal.II/base/quadrature_lib.h:99]: This is the same notation as in QGaussRadauChebychev,
but I'd be open to change both in this or a second PR.

Comment (jpthiele) [reply include/deal.II/base/quadrature_lib.h:102]: The notation here is the same as in QGaussRadauChebychev,
so I opted for consistency.

File: include/deal.II/base/quadrature_lib.h (hunk 3 above)
Comment (jpthiele) [reply include/deal.II/base/quadrature_lib.h:51]: Done

File: source/base/quadrature_lib.cc (hunk 4 above)
Comment (jpthiele) [reply source/base/quadrature_lib.cc]: Done

File: tests/base/quadrature_test.output [hunk 7]
```diff
@@ -83,6 +131,22 @@ DEAL:2d:subfaces::Quadrature no.17 is exact for polynomials of degree 5
...
+DEAL:2d:subfaces::Quadrature no.21 is exact for polynomials of degree 0
+DEAL:2d:subfaces::Quadrature no.22 is exact for polynomials of degree 0
+DEAL:2d:subfaces::Quadrature no.23 is exact for polynomials of degree 2
+DEAL:2d:subfaces::Quadrature no.24 is exact for polynomials of degree 2
+DEAL:2d:subfaces::Quadrature no.25 is exact for polynomials of degree 4
+DEAL:2d:subfaces::Quadrature no.26 is exact for polynomials of degree 4
+DEAL:2d:subfaces::Quadrature no.27 is exact for polynomials of degree 6
+DEAL:2d:subfaces::Quadrature no.28 is exact for polynomials of degree 6
+DEAL:2d:subfaces::Quadrature no.29 is exact for polynomials of degree 8
+DEAL:2d:subfaces::Quadrature no.30 is exact for polynomials of degree 8
```
Comment (peterrum) [inline tests/base/quadrature_test.output:143]: @jpthiele  Looks correct now!?

Comment (jpthiele) [reply tests/base/quadrature_test.output:143]: Yes, the quadrature formulas are indeed correct now.
Interestingly the final few digits seem to have had quite a large impact.

**Review by peterrum — APPROVED** [review]

File: include/deal.II/base/quadrature_lib.h [hunk 8]
```diff
@@ -47,6 +47,61 @@ class QGauss : public Quadrature<dim>
...
+ *
+ * This modification of the Gauss quadrature uses one of the two interval end
+ * points as well. Being exact for polynomials of degree <i>2n-2</i>, this
+ * formula is suboptimal by one degrees.
```
Comment (bangerth) [inline include/deal.II/base/quadrature_lib.h:55]: degrees -> degree

File: include/deal.II/base/quadrature_lib.h [hunk 9]
```diff
@@ -47,6 +47,61 @@ class QGauss : public Quadrature<dim>
...
+ *
+ * This formula is often used in the context of discontinuous Galerkin
+ * discretizations of ODEs and the temporal part of PDEs.
+ *
+ * The quadrature points are the left interval end point plus the roots of the
+ * polynomial <i>(P<sub>n-1</sub>(x)+P<sub>n</sub>(x))/(1+x)</i>
```
Comment (bangerth) [inline include/deal.II/base/quadrature_lib.h:61]: This works, but you can also use $latex-style$ formulas if you think that would be easier to read.

File: include/deal.II/base/quadrature_lib.h [hunk 10]
```diff
@@ -47,6 +47,61 @@ class QGauss : public Quadrature<dim>
...
+ * <i>(1-x<sub>i</sub>)/(n<sup>2</sup>(P<sub>n-1</sub>(x<sub>i</sub>))<sup>2</sup>)</i>.
+ *
+ * For the right Gauss-Radau formula the quadrature points are the points
+ * <i>1-x<sub>i</sub></i>, where <i>x<sub>i</sub></i> are the quadrature points
+ * of the left Gauss-Radau formula.
+ *
+ * @note This implementation only features the known exact roots up to
+ * n = 3 quadrature points.
+ *
+ * @sa https://mathworld.wolfram.com/RadauQuadrature.html @sa
```
Comment (bangerth) [inline include/deal.II/base/quadrature_lib.h:73]:
```suggestion
 * @seealso https://mathworld.wolfram.com/RadauQuadrature.html
```

File: include/deal.II/base/quadrature_lib.h (hunk 2 above)
Comment (bangerth) [reply include/deal.II/base/quadrature_lib.h:102]: That would be great! Feel free to leave it to a follow-up patch!

File: include/deal.II/base/quadrature_lib.h [hunk 11]
```diff
@@ -943,6 +998,8 @@ class QGaussPyramid : public Quadrature<dim>
 template <>
 QGauss<1>::QGauss(const unsigned int n);
 template <>
+QGaussRadau<1>::QGaussRadau(const unsigned int n, QGaussRadau<1>::EndPoint ep);
```
Comment (bangerth) [inline include/deal.II/base/quadrature_lib.h:1001]:
```suggestion
QGaussRadau<1>::QGaussRadau(const unsigned int n, const QGaussRadau<1>::EndPoint ep);
```

File: source/base/quadrature_lib.cc [hunk 12]
```diff
@@ -144,6 +144,232 @@ namespace internal
...
+            q_points[1] = 0.056262560536922135;
+            q_points[2] = 0.180240691736892389;
+            q_points[3] = 0.352624717113169672;
+            q_points[4] = 0.547153626330555420;
+            q_points[5] = 0.734210177215410598;
+            q_points[6] = 0.885320946839095790;
+            q_points[7] = 0.977520613561287499;
+            break;
+          default:
+            Assert(false, dealii::StandardExceptions::ExcNotImplemented());
```
Comment (bangerth) [inline source/base/quadrature_lib.cc:220]:
```suggestion
            Assert(false, ExcNotImplemented());
```

File: source/base/quadrature_lib.cc [hunk 13]
```diff
@@ -144,6 +144,232 @@ namespace internal
...
+  } // namespace QGaussRadau
+} // namespace internal
+
+#ifndef DOXYGEN
+template <>
+QGaussRadau<1>::QGaussRadau(const unsigned int n, EndPoint ep)
+  : Quadrature<1>(n)
+  , ep(ep)
+{
+  Assert(n > 0, ExcMessage("Need at least one point for quadrature rules"));
```
Comment (bangerth) [inline source/base/quadrature_lib.cc:360]:
```suggestion
  Assert(n > 0, ExcMessage("Need at least one point for quadrature rules."));
```

**Review by bangerth — COMMENTED** [review]
Very nice! Some comments, none should be much work!

**Comment by kronbichler** [comment]
To help me understand, I'm still wondering why the test previously passed for the left end point but not the right end point. It must be that we do not test the complete polynomial basis but only the monomials via https://github.com/dealii/dealii/blob/e8073dd8fabb7ec9b982aefe4efff7a0851e7782/tests/base/quadrature_test.cc#L78-L87 While the monomials form a basis, they form a pretty ill-conditioned one, so we might probably want to also look at `(1-x)` or `(0.5 - x)` for comparison just to be safe in terms of roundoff. If you agree @jpthiele I can post a follow-up later.

**Comment by jpthiele** [comment]
> To help me understand, I'm still wondering why the test previously passed for the left end point but not the right end point. 

In contrast to many other quadrature formulae the Gauss-Radau family is not symmetrical around 0 for [-1,1].
We did temporarily change the 1d test from x^i to (1-x)^i and the behaviour switched,
such that the left formula failed and the right formula passed.

@mathmerizing found a Julia package that computes Gauss formula of different families 
and it gave weights that differed to the previous results in the last three significant digits (i.e. around 14 identical ones).
There are two formulae listed for the weights 
and maybe the one with the derivative of the Legendre polynomials is a bit more accurate, 
if that was the one used in the other package.

File: include/deal.II/base/quadrature_lib.h (hunk 2 above)
Comment (jpthiele) [reply include/deal.II/base/quadrature_lib.h:102]: As there are various comments concerning the structure and commenting of both QGaussRadau 
and QGaussRadauChebychev I decided to apply all comments to QGaussRadau here.
After merging I will go over the QGaussRadauChebychev class 
and make all applicable changes to unify documentation and interface.

File: include/deal.II/base/quadrature_lib.h [hunk 14]
```diff
@@ -47,6 +47,71 @@ class QGauss : public Quadrature<dim>
...
+ *
+ * The quadrature points are the left interval end point plus the $n-1$
+ * roots of the polynomial
+ * \f[
+ *   \frac{P_{n-1}(x)+P_n(x)}{1+x}
+ * \f]
+ * where $P_{n-1}$ and $P_n$ are Legendre polynomials.
+ * The quadrature weights are
+ * \f[
+ *   w_0=\frac{2}{n^2}\quad\text{and}
```
Comment (jpthiele) [inline include/deal.II/base/quadrature_lib.h:68]: should be 

...\text{and}\quad
w_i=...

will fix that tomorrow.

File: include/deal.II/base/quadrature_lib.h [hunk 15]
```diff
@@ -47,6 +47,71 @@ class QGauss : public Quadrature<dim>
...
+ *   w_i=\quad\frac{1-x_i}{n^2(P_{n-1}(x_i))^2}\text{ for }i>0
+ * \f]
+ *
+ * For the right Gauss-Radau formula the quadrature points are the points
```
Comment (jpthiele) [inline include/deal.II/base/quadrature_lib.h:72]: Change to
... the quadrature points are $...$ and the weights are $\tilde{w}_i = w_{n-i-1}$, 
with $(x_i,w_i)$ as quadrature points and weights of the left Gauss-Radau formula.

File: include/deal.II/base/quadrature_lib.h (hunk 9 above)
Comment (jpthiele) [reply include/deal.II/base/quadrature_lib.h]: Yes it is easier to read. Changed it, thanks!

File: source/base/quadrature_lib.cc [hunk 16]
```diff
@@ -144,6 +144,234 @@ namespace internal
...
+      switch (end_point)
+        {
+          case ::dealii::QGaussRadau<1>::EndPoint::left:
+            return left_points;
+          case ::dealii::QGaussRadau<1>::EndPoint::right:
+            for (unsigned int i = 0; i < n; ++i)
+              {
+                points[n - i - 1] = 1. - left_points[i];
+              }
+            return points;
```
Comment (kronbichler) [inline source/base/quadrature_lib.cc:241]: Can you please move the declaration of `points` in here, i.e.,
```suggestion
          case ::dealii::QGaussRadau<1>::EndPoint::right:
            {
              std::vector<double> points(n);
              for (unsigned int i = 0; i < n; ++i)
                {
                  points[n - i - 1] = 1. - left_points[i];
                }
              return points;
            }
```
We aim to have the declaration near the place of use. As a side product, you avoid setting up the array when it's not needed.

**Review by kronbichler — APPROVED** [review]

**Review by kronbichler — APPROVED** [review]

**Comment by bangerth** [comment]
Yay, welcome to the authors of deal.II!

**Pull request merged**